File: clients/windows/src/Vertex.Shared/TunnelErrorReport.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Vertex.Shared;

/// <summary>
/// Categories the tunnel surfaces on a fatal disconnect. Mirror of Swift
/// <c>TunnelErrorKind</c> and Kotlin <c>TunnelErrorKind</c>. Wire string
/// values must stay byte-identical across platforms.
/// </summary>
[JsonConverter(typeof(TunnelErrorKindJsonConverter))]
public enum TunnelErrorKind
{
    Authentication,
    IdentityRejected,
    DiscoveryTimeout,
    JoinTimeout,
    Configuration,
    Unknown,
}

/// <summary>
/// Last fatal error from the tunnel. Persisted by the Service and read by
/// the App when status flips back to Disconnected. Field names match the
/// Kotlin <c>TunnelErrorReport</c> wire format (Swift's variant adds a
/// <c>keychainLocked</c> kind that doesn't apply on Windows).
/// </summary>
public sealed record TunnelErrorReport(
    [property: JsonPropertyName("kind")]                                                                            TunnelErrorKind Kind,
    [property: JsonPropertyName("detail")]                                                                          string          Detail           = "",
    [property: JsonPropertyName("timestampEpochMs"), JsonNumberHandling(JsonNumberHandling.AllowReadingFromString)] long            TimestampEpochMs = 0)
{
    public string UserMessage => Kind switch
    {
        TunnelErrorKind.Authentication =>
            $"Authentication failed. Check Client name and Password in Settings → Identity. ({Detail})",
        TunnelErrorKind.IdentityRejected =>
            $"The exit rejected this device's identity ({Detail}). Ask admin to reset TOFU for this device on the exit, then reconnect.",
        TunnelErrorKind.DiscoveryTimeout =>
            $"Exit \"{Detail}\" is unreachable. Check Edge selection in Settings or try a different exit.",
        TunnelErrorKind.JoinTimeout =>
            $"Exit \"{Detail}\" didn't respond to join. The exit may be down, or your Client name is not authorized.",
        TunnelErrorKind.Configuration =>
            $"Configuration error: {Detail}",
        TunnelErrorKind.Unknown =>
            string.IsNullOrEmpty(Detail) ? "Connection failed." : Detail,
        _ => "Connection failed.",
    };
}

internal sealed class TunnelErrorKindJsonConverter : JsonConverter<TunnelErrorKind>
{
    public override TunnelErrorKind Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        reader.GetString() switch
        {
            "authentication"    => TunnelErrorKind.Authentication,
            "identity_rejected" => TunnelErrorKind.IdentityRejected,
            "discovery_timeout" => TunnelErrorKind.DiscoveryTimeout,
            "join_timeout"      => TunnelErrorKind.JoinTimeout,
            "configuration"     => TunnelErrorKind.Configuration,
            "unknown"           => TunnelErrorKind.Unknown,
            var s => throw new JsonException($"Unknown TunnelErrorKind '{s}'"),
        };

    public override void Write(Utf8JsonWriter writer, TunnelErrorKind value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value switch
        {
            TunnelErrorKind.Authentication   => "authentication",
            TunnelErrorKind.IdentityRejected => "identity_rejected",
            TunnelErrorKind.DiscoveryTimeout => "discovery_timeout",
            TunnelErrorKind.JoinTimeout      => "join_timeout",
            TunnelErrorKind.Configuration    => "configuration",
            TunnelErrorKind.Unknown          => "unknown",
            _ => throw new JsonException($"Unhandled TunnelErrorKind {value}"),
        });
}

File: clients/android/Vertex/core/src/main/kotlin/ru/vertices/android/core/ipc/TunnelErrorReport.kt
package ru.vertices.android.core.ipc

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Error categories the tunnel surfaces to the UI on a fatal disconnect.
 * Mirror of Swift `TunnelErrorKind`.
 */
@Serializable
enum class TunnelErrorKind {
    @SerialName("authentication")      AUTHENTICATION,
    @SerialName("identity_rejected")   IDENTITY_REJECTED,
    @SerialName("discovery_timeout")   DISCOVERY_TIMEOUT,
    @SerialName("join_timeout")        JOIN_TIMEOUT,
    @SerialName("configuration")       CONFIGURATION,
    @SerialName("unknown")              UNKNOWN,
}

/**
 * Last fatal error from the tunnel. Persisted to disk by VpnService and read
 * by the host app when status flips back to DISCONNECTED.
 */
@Serializable
data class TunnelErrorReport(
    val kind: TunnelErrorKind,
    val detail: String = "",
    val timestampEpochMs: Long = System.currentTimeMillis(),
) {
    /** Localized user-facing message. Equivalent to Swift `userMessage`. */
    val userMessage: String
        get() = when (kind) {
            TunnelErrorKind.AUTHENTICATION ->
                "Authentication failed. Check Client name and Password in Settings → Identity. ($detail)"
            TunnelErrorKind.IDENTITY_REJECTED ->
                "The exit rejected this device's identity ($detail). Ask admin to reset TOFU for this device on the exit, then reconnect."
            TunnelErrorKind.DISCOVERY_TIMEOUT ->
                "Exit \"$detail\" is unreachable. Check Edge selection in Settings or try a different exit."
            TunnelErrorKind.JOIN_TIMEOUT ->
                "Exit \"$detail\" didn't respond to join. The exit may be down, or your Client name is not authorized."
            TunnelErrorKind.CONFIGURATION ->
                "Configuration error: $detail"
            TunnelErrorKind.UNKNOWN ->
                if (detail.isEmpty()) "Connection failed." else detail
        }
}
